chain/chain/src/missing_chunks.rs

```diff
@@ -0,0 +1,292 @@
+        // has gone horribly wrong, in which case these HashMaps will be lost anyways.
+        if self.blocks_missing_chunks.len() >= MAX_BLOCKS_MISSING_CHUNKS {
+            warn!(target: "chunks", "Not recording block with hash {} even though it is missing chunks. The missing chunks pool is full.", block_hash);
+            return;
+        }
+
+        for chunk_hash in missing_chunks.iter().cloned() {
+            let blocks_for_chunk =
+                self.missing_chunks.entry(chunk_hash).or_insert_with(HashSet::new);
+            blocks_for_chunk.insert(block_hash);
+        }
+
+        // Convert to HashSet
+        let missing_chunks = missing_chunks.into_iter().collect();
+        match self.blocks_missing_chunks.entry(block_hash) {
+            hash_map::Entry::Vacant(entry) => {
+                entry.insert(missing_chunks);
+            }
+            // The Occupied case should never happen since we know
+            // all the missing chunks for a block the first time we receive it,
+            // and we should not call `add_block_with_missing_chunks` again after
+            // we know a block is missing chunks.
+            hash_map::Entry::Occupied(mut entry) => {
+                let previous_chunks = entry.insert(missing_chunks);
+                warn!(target: "chunks", "Block with hash {} was already missing chunks {:?}.", block_hash, previous_chunks);
+            }
+        }
+
+        let height = block.height();
+        let blocks_at_height = self.height_idx.entry(height).or_insert_with(HashSet::new);
+        blocks_at_height.insert(block_hash);
+        self.blocks_waiting_for_chunks.insert(block_hash, block);
+    }
+
+    pub fn accept_chunk(&mut self, chunk_hash: &ChunkHash, height_included: BlockHeight) {
+        let block_hashes = self.missing_chunks.remove(chunk_hash).unwrap_or_else(HashSet::new);
+        for block_hash in block_hashes {
+            match self.blocks_missing_chunks.entry(block_hash) {
+                hash_map::Entry::Occupied(mut missing_chunks_entry) => {
+                    let missing_chunks = missing_chunks_entry.get_mut();
+                    missing_chunks.remove(chunk_hash);
+                    if missing_chunks.is_empty() {
+                        // No more missing chunks!
+                        missing_chunks_entry.remove_entry();
+                        self.mark_block_as_ready(&block_hash);
+                    }
+                }
+                hash_map::Entry::Vacant(_) => {
+                    warn!(target: "chunks", "Invalid MissingChunksPool state. Block with hash {} was still a value of the missing_chunks map, but not present in the blocks_missing_chunks map", block_hash);
+                    self.mark_block_as_ready(&block_hash);
+                }
+            }
+        }
+        // Chunks are accepted in order of height (because they rely on knowing the prev_block),
```

Comment:
The reasoning in the parenthesis does not imply that chunks are accepted in order of height. Forks do exist

Reply:
The only purpose of this is to prune old blocks from the data structure (i.e. prevent forever unavailable chunks from blowing this up). I'll re-think how to do this in a safe way since your other comment about `height_included` is about this same piece of logic.

The consensus protocol has finalization, I know that. What if instead I pass in the last finalized height here instead of the last added block? I think I could do that at the end of `on_block_accepted` for example.

Reply:
An easy solution is to prune heights below `Chain::tail()` when some block is accepted.

Reply:
I changed the pruning to use `Chain::tail` in https://github.com/near/nearcore/pull/3710/commits/54c6b603d2a0452aee8a7e16f14e7d44e2068526

I made it happen in `receive_block` instead of on block acceptance because we already had a call to `tail` there, so I could use that result again. Of course it could mean we are pruning the `MissingChunksPool` more often than necessary, but I thought that is ok because pruning on an already pruned pool should be a no-op that takes essentially no time. Whereas calling `tail` looks like a DB read and thus could be more expensive to repeat that in some other part of the code.

But let me know what you think.